tests/unit-tests/test-class-admin.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+class Sensei_Class_Admin_Test extends WP_UnitTestCase {
+
+	/**
+	 * Constructor function
+	 */
+	public function __construct() { // phpcs:ignore Generic.CodeAnalysis.UselessOverridingMethod.Found
+		parent::__construct();
+	}
+
+	/**
+	 * Setup function
+	 *
+	 * This function sets up the lessons, quizes and their questions. This function runs before
+	 * every single test in this class
+	 *
+	 * @return void
+	 */
+	public function setup() {
+		parent::setup();
+
+		$this->factory = new Sensei_Factory();
+	}//end setup()
+
+	public function tearDown() {
+		parent::tearDown();
+		$this->factory->tearDown();
+	}
+
+	/**
+	 * Testing the admin class to make sure it is loaded
+	 *
+	 * @return void
+	 */
+	public function testClassInstance() {
+		// test if the class exists
+		$this->assertTrue( class_exists( 'WooThemes_Sensei_Admin' ), 'Sensei Admin class does not exist' );
+	} // end testClassInstance
+
+	/**
+	 * Setup function for duplicate course
+	 *
+	 * This function mocks the redirect method, create and set the user, create a course
+	 * with some lessons and set the needed params.
+	 *
+	 * @param $qty_lessons How many lessons to create.
+	 * @return array       Created course and list of lessons id.
+	 */
+	private function duplicate_course_with_lessons_setup( $qty_lessons ) {
+		// Mock the safe_redirect method
+		Sensei()->admin = $this->getMockBuilder( 'WooThemes_Sensei_Admin' )
+			->setMethods( [ 'safe_redirect' ] )
+			->getMock();
+
+		Sensei()->admin->expects( $this->once() )
+			->method( 'safe_redirect' );
+
+		// Create and set current user as teacher
+		$admin_user = $this->factory->user->create_and_get(
+			array(
+				'user_login' => 'admin_user',
+				'user_pass'  => null,
+				'role'       => 'teacher',
+			)
+		);
+		wp_set_current_user( $admin_user->ID );
+
+		// Create one course and add some lessons
+		$course_id   = $this->factory->course->create();
+		$lessons_ids = $this->factory->lesson->create_many( $qty_lessons );
+
+		foreach ( $lessons_ids as $lesson_id ) {
+			add_post_meta( $lesson_id, '_lesson_course', $course_id );
+		}
+
+		// Create nonce
+		$duplicate_nonce = wp_create_nonce( 'duplicate_course_with_lessons_' . $course_id );
+
+		// Set request and get params
+		$_REQUEST['_wpnonce'] = $duplicate_nonce;
+		$_GET['post']         = $course_id;
+
+		return array(
+			'course_id'   => $course_id,
+			'lessons_ids' => $lessons_ids,
+		);
+	}
+
+	/**
+	 * Test duplicate courses with lessons.
+	 *
+	 * @covers WooThemes_Sensei_Admin::duplicate_course_with_lessons_action
```

Comment:
This test ends up covering more than just this function. It drills down into `duplicate_content` and then probably some other intermediary functions before finally getting to the `private` functions we are ultimately writing these tests for 😅. We should list all those functions here.